### PR TITLE
Typeahead

### PR DESCRIPTION
**Pull request by JeffreyMFarley**
Typeahead!

## Testing

- Testing remains at 100%

## Review

- @sephcoster @kurzn 

#### One character typed
screen shot 2017-07-07 at 4 07 40 pm

#### Two characters typed
screen shot 2017-07-07 at 4 08 01 pm

#### Less than 5 results
screen shot 2017-07-07 at 4 08 27 pm

#### Arrow key selecting an option
screen shot 2017-07-07 at 4 08 45 pm

#### Arrow key selecting one item, mouse hover over another
screen shot 2017-07-07 at 4 09 01 pm

#### After the user has selected an option
screen shot 2017-07-07 at 4 09 18 pm

## To Dos (in another PR)

- [ ] Support for asynchronous callbacks (like when calling the suggest endpoint)
- [ ] Enter and Tab will work differently when the typeahead is used with the search suggester.  Currently, the user _must_ select an option. Free form values are not allowed
- [ ] May need an option to not clear the results after selecting

## Checklist

* [x] Changes are limited to a single goal (no scope creep)
* [x] Code can be automatically merged (no conflicts)
* [x] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [x] Passes all existing automated tests
* [x] New functions include new tests
* [x] New functions are documented (with a description, list of inputs, and expected output)
* [x] Placeholder code is flagged
* [x] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

File: src/Filters/Issue.jsx [hunk 1]
```diff
@@ -32,11 +47,65 @@ export class Issue extends React.Component {
...
+      key: x.key,
+      normalized: normalize(x.key)
+    }
+  })
+
   return {
-    options: state.aggs.issue
+    options,
+    forTypeahead
   }
```
Comment (sephcoster) [inline src/Filters/Issue.jsx]: Is the plan for the ToDo on line 92 to add a MapDispatchToProps with a variant on our existing `FILTER_CHANGE` actions? It occurs to me we may also need a `TOGGLE_PARENT_FILTER` action as well to accomplish this.

File: src/Typeahead/index.jsx [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+import React from 'react'
+import PropTypes from 'prop-types'
+import Selector from './Selector'
+import * as keys from '../constants'
+import './Typeahead.less'
+
+// ----------------------------------------------------------------------------
+// State Machine
+
+// TODO: Use with onInputChangeAsynch
```
Comment (sephcoster) [inline src/Typeahead/index.jsx]: This is a cool way to approach all the edge cases / possible options for this.

File: src/Typeahead/index.jsx [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    return (
+      <section className="typeahead"
+               onBlur={this._onBlur}
+               onFocus={this._onFocus}>
+        <input type="text"
+               disabled={this.props.disabled}
+               onChange={this._valueUpdated}
+               onKeyDown={this._onKeyDown}
+               placeholder={this.props.placeholder}
+               value={this.state.inputValue}
```
Comment (sephcoster) [inline src/Typeahead/index.jsx]: Maybe here's a better place to set the top-level stuff and pass down so you can change it per distinct type-ahead instance (search Dispatch vs. Issue dispatch, etc)

File: src/Filters/Issue.jsx (hunk 1 above)
Comment (JeffreyMFarley) [inline src/Filters/Issue.jsx]: I knew we had to refactor the parent/child selection logic, so I figured I would handle that TODO when I update the logic

File: src/Typeahead/index.jsx (hunk 3 above)
Comment (JeffreyMFarley) [inline src/Typeahead/index.jsx]: Yeah, if this code needs to be extended we can look into that

**Pull request merged**